### PR TITLE
Issue-6089 Set go.property() textures and fonts for `gui`component 

### LINKED ISSUE #6089: Set go.property() texture and font for `gui` component
**Is your feature request related to a problem? Please describe (REQUIRED):**
Sometimes it's useful to load only one font for `gui`, and then apply another one from go, property to make localisation. The same thing with GUI and textures. When user wants to change skins of the main gui without loading all the skins simultaneously.

**Describe the solution you'd like (REQUIRED):**

At the moment we have 
```
go.property("my_atlas", resource.atlas("/atlas.atlas"))
function init(self)
  go.set("#sprite", "image", self.my_atlas)
end
```
It would be useful to have the same for the *.gui  component:

```
go.property("my_atlas", resource.atlas("/atlas.atlas"))
function init(self)
  go.set("#gui_component", "texture", self.my_atlas)
end
```

**Describe alternatives you've considered (REQUIRED):**
no

### PR DESCRIPTION
**Pull request by AGulev**
Added a possibility to set `go.property()` textures and fonts from for GUI components.

```lua
go.property("font", resource.font())
go.property("atlas", resource.atlas())

function init(self)
	local system_font = go.get("#gui", "fonts", {key = "system_font"})
	go.set("#gui", "fonts", self.font, {key = "system_font"}) --replace 
        go.set("#gui", "fonts", self.font, {key = "system_font2"}) --set new 
	-- reset old font
	go.set("#gui", "fonts", system_font, {key = "system_font"})
	

	local old_atlas = go.get("#gui", "textures", {key = "my_atlas"})
	go.set("#gui", "textures", self.atlas, {key = "my_atlas"}) -- replace
        go.set("#gui", "textures", self.atlas, {key = "my_atlas2"}) -- set new
        -- reset old atlas
	go.set("#gui", "fonts", old_atlas, {key = "my_atlas"})
end
```

### CODE REVIEW

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 1]
```diff
@@ -588,10 +588,10 @@ namespace dmGameSystem
...
         scene_params.m_MaxNodes = scene_desc->m_MaxNodes;
-        scene_params.m_MaxAnimations = 1024;
+        scene_params.m_MaxAnimations = ANIMATIONS_MAX;
         scene_params.m_UserData = gui_component;
-        scene_params.m_MaxFonts = 64;
-        scene_params.m_MaxTextures = 128;
+        scene_params.m_MaxFonts = FONTS_MAX;
+        scene_params.m_MaxTextures = TEXTURES_MAX;
         scene_params.m_MaxParticlefx = gui_world->m_MaxParticleFXCount;
         scene_params.m_MaxSpineScenes = 128;
```
Comment (AGulev) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:596]: I didn't touch Spine here because I know @JCash is going to touch it very soon.

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 2]
```diff
@@ -2116,9 +2122,24 @@ namespace dmGameSystem
...
+        dmhash_t set_property = params.m_PropertyId;
+        if (set_property == PROP_MATERIAL) {
             return SetResourceProperty(dmGameObject::GetFactory(params.m_Instance), params.m_Value, MATERIAL_EXT_HASH, (void**)&gui_component->m_Material);
         }
+        else if (set_property == PROP_FONT) {
+            dmRender::HFontMap font;
+            memset(&font, 0, sizeof(dmRender::HFontMap));
+            dmGameObject::PropertyResult res = SetResourceProperty(dmGameObject::GetFactory(params.m_Instance), params.m_Value, FONT_EXT_HASH, (void**)&font);
+            dmGui::HScene scene = gui_component->m_Scene;
+            dmGui::Result r = dmGui::AddFont(scene, params.m_Value.m_Hash, (void*) font, params.m_Value.m_Hash);
```
Comment (AGulev) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:2134]: We want to operate these fonts as a regular font, so we add it in GUI fonts using resource path hash as name, because we doesn't have any other name. I think developer knows himself what font in what property he has.

So it may be used like this:

`font_prop.script`
```
go.property("font", resource.font())
function init(self)
	go.set("#gui", "font", self.font)
	msg.post("#font", "change", {font = self.font})
end
```

`font_prop.gui_script`
```
function on_message(self, message_id, message, sender)
	if message_id == hash("change") then
		gui.set_font(gui.get_node("text"), message.font)
	end
end
```

File: engine/gamesys/src/gamesys/components/comp_gui_private.h [hunk 3]
```diff
@@ -23,6 +23,10 @@
 
 namespace dmGameSystem
 {
+    static const uint32_t FONTS_MAX         = 64;
+    static const uint32_t TEXTURES_MAX      = 128;
+    static const uint32_t ANIMATIONS_MAX    = 1024;
```
Comment (AGulev) [inline engine/gamesys/src/gamesys/components/comp_gui_private.h:28]: I need these values to know how many pointers to resource property we can use

File: engine/gui/src/gui.h [hunk 4]
```diff
@@ -653,18 +653,18 @@ namespace dmGui
      * Adds a font with the specified name to the scene.
      * @note Any nodes connected to the same font_name will also be connected to the new font. This makes this function O(n), where n is #nodes.
      * @param scene Scene to add texture to
-     * @param font_name Name of the font that will be used in the gui scripts
+     * @param font_name_hash Hash of the font name that will be used in the gui scripts
      * @param font The font to add
      * @return Outcome of the operation
      */
-    Result AddFont(HScene scene, const char* font_name, void* font, dmhash_t path_hash);
+    Result AddFont(HScene scene, dmhash_t font_name_hash, void* font, dmhash_t path_hash);
```
Comment (AGulev) [inline engine/gui/src/gui.h:660]: I think it e sense to have one function that uses `dmhash_t` instead of creation one more funciton

**Review by AGulev — COMMENTED** [review]
only fonts for now

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 5]
```diff
@@ -2116,9 +2122,24 @@ namespace dmGameSystem
...
     dmGameObject::PropertyResult CompGuiSetProperty(const dmGameObject::ComponentSetPropertyParams& params) {
         GuiComponent* gui_component = (GuiComponent*)*params.m_UserData;
-        if (params.m_PropertyId == PROP_MATERIAL) {
+        dmhash_t set_property = params.m_PropertyId;
+        if (set_property == PROP_MATERIAL) {
             return SetResourceProperty(dmGameObject::GetFactory(params.m_Instance), params.m_Value, MATERIAL_EXT_HASH, (void**)&gui_component->m_Material);
         }
+        else if (set_property == PROP_FONT) {
+            dmRender::HFontMap font;
+            memset(&font, 0, sizeof(dmRender::HFontMap));
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:2131]: We generally don't do defensive programming.
Why is this necessary here?

You should check the result from the SetResourceProperty

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 2 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:2134]: It's not really a "set" here, but an "add"

I think this api leaves too much ambiguity. E.g.:
> go.set("#gui", "font", self.font1)
> go.set("#gui", "font", self.font2)

Which font is now "set" ?

I'm not sure what's the best way to do it is here, but as a worst case, we should be able to use a new message:
> msg.post(gui_url, "add_resource", { alias = alias, resource = hash, type = ".fontc" }

(probably possible to detect extension programattically fromt he resource system)

Or possibly separate messages: "add_font"/"add_texture"

There's also the issue of "removing" references (if we need it)
E.g. clearing references.

File: engine/gamesys/src/gamesys/components/comp_gui_private.h [hunk 6]
```diff
@@ -34,6 +38,8 @@ namespace dmGameSystem
         uint16_t                m_ComponentIndex;
         uint8_t                 m_Enabled : 1;
         uint8_t                 m_AddedToUpdate : 1;
+        void*                   m_resourcePropertyPointers[FONTS_MAX + TEXTURES_MAX];
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui_private.h:41]: I think you should just use a dmArray<void*> instead.

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h [hunk 7]
```diff
@@ -250,10 +252,12 @@ namespace dmGameObject
      * @struct
      * @name PropertyOptions
      * @member m_Index [type:int32_t] The index of the property to set, only applicable if property is array.
+     * @member m_Key [type:char*] The key of the property to set, only applicable if property is hashtable.
      */
     struct PropertyOptions
     {
         int32_t m_Index;
+        dmhash_t m_Key;
     };
```
Comment (JCash) [inline engine/gameobject/src/dmsdk/gameobject/gameobject.h:261]: Given that a property _might_ support both ways of indexing, we need a way to distinguish between the two.

E.g.:
```
    struct PropertyOptions
    {
        union {
            dmhash_t m_Key;
            int32_t m_Index;
        }; // It now takes 8 bytes and not 16

        uint8_t m_HasKey:1: // it takes 9 bytes with this
    };
```

File: engine/gameobject/src/gameobject/gameobject_script.cpp [hunk 8]
```diff
@@ -748,7 +750,25 @@ namespace dmGameObject
...
+                }
+                else if (dmScript::IsHash(L, -1))
+                {
+                    property_options.m_Key = dmScript::CheckHash(L, -1);
+                }
+                else
+                {
+                    return luaL_error(L, "Invalid key argument in options table. It should be string or hash.");
+                }
+            }
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject_script.cpp:771]: I think we should handle the fact that this code kan use index+key at the same time. I don't think we want that.

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 9]
```diff
@@ -604,6 +604,9 @@ namespace dmGameSystem
         gui_component->m_Scene = dmGui::NewScene(scene_resource->m_GuiContext, &scene_params);
         dmGui::HScene scene = gui_component->m_Scene;
 
+        gui_component->m_resourcePropertyPointers.SetCapacity(scene_params.m_MaxTextures + scene_params.m_MaxFonts);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:607]: My point with the dmArray implementation is that it is a dynamic array.
There's really rarely any need to preallocate this data, as most users will never use this feature.

Instead, call OffsetCapacity() if the buffer is full when we want to add another item. We can still have an upper limit if we want to.

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 10]
```diff
@@ -2116,9 +2126,25 @@ namespace dmGameSystem
...
             return SetResourceProperty(dmGameObject::GetFactory(params.m_Instance), params.m_Value, MATERIAL_EXT_HASH, (void**)&gui_component->m_Material);
         }
+        else if (set_property == PROP_FONTS) {
+            if (!params.m_Options.m_Key) {
+                return dmGameObject::PROPERTY_RESULT_INVALID_KEY;
+            }
+            dmRender::HFontMap font = 0x0;
+            dmGameObject::PropertyResult res = SetResourceProperty(dmGameObject::GetFactory(params.m_Instance), params.m_Value, FONT_EXT_HASH, (void**)&font);
+            dmGui::HScene scene = gui_component->m_Scene;
+            dmGui::Result r = dmGui::AddFont(scene, params.m_Options.m_Key, (void*) font, params.m_Value.m_Hash);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:2140]: How do we want to deal with "resetting" this property?
E.g. we have "sprite.reset_constant()" in order to go back to the previous version

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 11]
```diff
@@ -2116,9 +2126,25 @@ namespace dmGameSystem
...
+            gui_component->m_resourcePropertyPointers.Push(font);
+            if (r != dmGui::RESULT_OK) {
+                dmLogError("Unable to set font property");
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:2143]: More info: E.g. Which font was it? Do we have the component name?

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 12]
```diff
@@ -2116,9 +2126,25 @@ namespace dmGameSystem
...
+                return dmGameObject::PROPERTY_RESULT_BUFFER_OVERFLOW;
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:2144]: Don't we want to decrement the reference here?
I.e. I don't think we should keep the resource if setting it failed

File: engine/gamesys/src/gamesys/components/comp_gui_private.h [hunk 13]
```diff
@@ -34,6 +34,7 @@ namespace dmGameSystem
         uint16_t                m_ComponentIndex;
         uint8_t                 m_Enabled : 1;
         uint8_t                 m_AddedToUpdate : 1;
+        dmArray<void*>          m_resourcePropertyPointers;
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui_private.h:37]: naming convention -> m_ResourcePropertyPointers

File: engine/gui/src/test/test_gui.cpp [hunk 14]
```diff
@@ -2813,8 +2813,8 @@ TEST_F(dmGuiTest, DeltaTime)
 
 TEST_F(dmGuiTest, Bug352)
 {
-    dmGui::AddFont(m_Scene, "big_score", 0, 0);
-    dmGui::AddFont(m_Scene, "score", 0, 0);
+    dmGui::AddFont(m_Scene, dmHashString64("big_score"), 0, 0);
+    dmGui::AddFont(m_Scene, dmHashString64("score"), 0, 0);
```
Comment (JCash) [inline engine/gui/src/test/test_gui.cpp:2817]: I want a lua test as well for the go.set() with keys

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/gameobject/src/gameobject/gameobject_script.cpp (hunk 8 above)
Comment (AGulev) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:771]: so we should throw Lua error if there are key AND index in this table at the same time?

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 9 above)
Comment (AGulev) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:607]: 👍makes sense, I'll fix that

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 10 above)
Comment (AGulev) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:2140]: I think it's a good "nice to have" feature we would like to design and implement in the future because we wanna use it not only here, but as replacement for deprecated `sprite.reset_constant()` function.

At this moment, in this particular case with fonts and textures, it's easy to workaround by getting value before set, and then set it later again (reset).

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 11 above)
Comment (AGulev) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:2143]: 👍

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 12 above)
Comment (AGulev) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:2144]: Good point!

File: engine/gamesys/src/gamesys/components/comp_gui_private.h (hunk 13 above)
Comment (AGulev) [reply engine/gamesys/src/gamesys/components/comp_gui_private.h:37]: 👍

File: engine/gui/src/test/test_gui.cpp (hunk 14 above)
Comment (AGulev) [reply engine/gui/src/test/test_gui.cpp:2817]: yes, it's important for sure. I'll do

File: engine/gameobject/src/gameobject/gameobject_script.cpp (hunk 8 above)
Comment (AGulev) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:818]: done

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 10 above)
Comment (AGulev) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:2146]:
```
	local system_font = go.get("#gui", "fonts", {key = "system_font"})
	go.set("#gui", "fonts", self.font, {key = "system_font"})
	--reset
	go.set("#gui", "fonts", system_font, {key = "system_font"})
```

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h [hunk 15]
```diff
@@ -250,10 +252,12 @@ namespace dmGameObject
      * @struct
      * @name PropertyOptions
      * @member m_Index [type:int32_t] The index of the property to set, only applicable if property is array.
+     * @member m_Key [type:char*] The key of the property to set, only applicable if property is hashtable.
      */
     struct PropertyOptions
     {
         int32_t m_Index;
+        const char* m_Key;
```
[the remark this answers is not in the document]
Comment (AGulev) [reply engine/gameobject/src/dmsdk/gameobject/gameobject.h]: Done! thx!

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h (hunk 7 above)
Comment (AGulev) [reply engine/gameobject/src/dmsdk/gameobject/gameobject.h]: Done! thx!

File: engine/engine/src/test/script_props/spawn.script [hunk 16]
```diff
@@ -20,5 +22,5 @@ function init(self, config)
     assert(self.quat.y == 9)
     assert(self.quat.z == 10)
     assert(self.quat.w == 11)
-    msg.post("main:/main#script", "done")
```
Comment (AGulev) [inline engine/engine/src/test/script_props/spawn.script:23]: we call it 3 times per collection, so some tests have never done. I added "done" counter to use tests for similiar features in one place but in different GOs and scripts

File: engine/engine/src/test/script_props/uniform_property.script [hunk 17]
```diff
@@ -66,8 +68,8 @@ function init(self, config)
 
     --Test setting a single element to a constant
     assert_not_error(function() go.set("#sprite", "uniform_array.y", 7, { index = 15}) end)
-    assert_number(go.get("#sprite", "uniform_array.y", { index = 17 }), 7)
```
Comment (AGulev) [inline engine/engine/src/test/script_props/uniform_property.script:69]: one of the tests which is never done. we set element 15 but try to read element 17. I fixed that

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h [hunk 18]
```diff
@@ -243,17 +245,42 @@ namespace dmGameObject
...
+     * @member dmGameObject::PROP_VALUE_VAR
+     * @member dmGameObject::PROP_VALUE_ARRAY
+     * @member dmGameObject::PROP_VALUE_HASHTABLE
+     */
+    enum PropertyValueType
+    {
+        PROP_VALUE_VARIABLE = 0,
+        PROP_VALUE_ARRAY = 1,
+        PROP_VALUE_HASHTABLE = 2,    
+    };
```
Comment (AGulev) [inline engine/gameobject/src/dmsdk/gameobject/gameobject.h:263]: We wanna know property type to provide better errors handling

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h [hunk 19]
```diff
@@ -326,7 +353,7 @@ namespace dmGameObject
         float* m_ValuePtr;
         bool m_ReadOnly;
         uint8_t m_ArraySize : 7;
-        uint8_t m_IsArray   : 1;
+        uint8_t m_ValueType   : 2;
```
Comment (AGulev) [inline engine/gameobject/src/dmsdk/gameobject/gameobject.h:356]: 1 bit isn't enough anymore

**Comment by AGulev** [comment]
@JCash 
I'm not sure why windows tests are failed. I spent a some time today trying to figure that out, tomorrow I'll try on windows machine again. Other that that isn't ready for review.

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h [hunk 20]
```diff
@@ -325,8 +351,7 @@ namespace dmGameObject
         PropertyVar m_Variant;
         float* m_ValuePtr;
         bool m_ReadOnly;
-        uint8_t m_ArraySize : 7;
```
Comment (AGulev) [inline engine/gameobject/src/dmsdk/gameobject/gameobject.h:328]: this value is never used

**Comment by AGulev** [comment]
@JCash it's green, finally

File: engine/gameobject/src/gameobject/gameobject_script.cpp [hunk 21]
```diff
@@ -595,9 +596,32 @@ namespace dmGameObject
...
+            if (!lua_isnil(L, -1))
+            {
+                if (lua_isstring(L, -1))
+                {
+                    property_options.m_Key = dmHashString64(lua_tostring(L, -1));
+                }
+                else if (dmScript::IsHash(L, -1))
+                {
+                    property_options.m_Key = dmScript::CheckHash(L, -1);
+                }
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject_script.cpp:609]: I think it's enough with "key = dmScript::CheckHashOrString(L, -1)" as it also handles error reporting

File: engine/gameobject/src/gameobject/gameobject_script.cpp [hunk 22]
```diff
@@ -726,16 +778,40 @@ namespace dmGameObject
...
+                else
+                {
+                    return luaL_error(L, "Invalid key argument in options table. It should be string or hash.");
+                }
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject_script.cpp:803]: Same as above, CheckHashOrString() should suffice

File: engine/gameobject/src/gameobject/gameobject_script.cpp [hunk 23]
```diff
@@ -726,16 +778,40 @@ namespace dmGameObject
...
+            }
+            lua_pop(L, 1);
+
             lua_getfield(L, -1, "index");
             if (!lua_isnil(L, -1)) // make it optional
-            {
+            {   
+                if (property_options.m_HasKey)
+                {
+                    return luaL_error(L, "Options table should contain either key or index value.");
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject_script.cpp:813]: I think the message is a bit ambiguous.
Perhaps "Options table cannot contain both 'key' and 'index'" ?

File: engine/gamesys/src/gamesys/gamesys_private.cpp [hunk 24]
```diff
@@ -83,8 +83,7 @@ namespace dmGameSystem
                 value = &value[value_index];
             }
 
-            out_desc.m_ArraySize = constant_array_size;
-            out_desc.m_IsArray   = constant_array_size > 1;
+            out_desc.m_ValueType   = (constant_array_size > 1) ? dmGameObject::PROP_VALUE_ARRAY : dmGameObject::PROP_VALUE_VARIABLE;
```
Comment (JCash) [inline engine/gamesys/src/gamesys/gamesys_private.cpp:86]: Now when I look at it, I wonder: Do we really need/want the "PROP_VALUE_VARIABLE"?
A variable can just as easily be represented by an array of 1 element.
The only special case I can think of is get/set of a variable with no "index" specified.

But, we can cover that case using 0 as default index if there is no "index" specified.


That would reduce the 3 types down to 2, which arguably is a lot easier to wrap your head around.

What do you think?

File: engine/engine/src/test/script_props/test_done.lua [hunk 25]
```diff
@@ -0,0 +1,12 @@
+local M = {}
+
+M.test_count = 3
```
[the remark this answers is not in the document]
Comment (AGulev) [reply engine/engine/src/test/script_props/test_done.lua]: 👍

File: engine/gameobject/src/gameobject/gameobject_script.cpp (hunk 21 above)
Comment (AGulev) [reply engine/gameobject/src/gameobject/gameobject_script.cpp]: soooo much better! thx

File: engine/gameobject/src/gameobject/gameobject_script.cpp (hunk 22 above)
Comment (AGulev) [reply engine/gameobject/src/gameobject/gameobject_script.cpp]: 👍

File: engine/gameobject/src/gameobject/gameobject_script.cpp (hunk 23 above)
Comment (AGulev) [reply engine/gameobject/src/gameobject/gameobject_script.cpp]: 👍

File: engine/gamesys/src/gamesys/gamesys_private.cpp (hunk 24 above)
Comment (AGulev) [reply engine/gamesys/src/gamesys/gamesys_private.cpp]: done

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 26]
```diff
@@ -659,9 +659,16 @@ namespace dmGameSystem
...
                 if (gui_component->m_Material) {
-                    dmResource::Release(dmGameObject::GetFactory(params.m_Instance), gui_component->m_Material);
+                    dmResource::Release(factory, gui_component->m_Material);
                 }
+                for (uint32_t i = 0; i < gui_component->m_ResourcePropertyPointers.Size(); ++i) {
+                    if (gui_component->m_ResourcePropertyPointers[i]) {
+                        dmResource::Release(factory, gui_component->m_ResourcePropertyPointers[i]);
+                    }
+                }
+                gui_component->m_ResourcePropertyPointers.SetSize(0);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:671]: Fyi, no real need to set size here, as the memory is release two lines down

File: engine/gamesys/src/gamesys/gamesys_private.cpp [hunk 27]
```diff
@@ -83,8 +83,7 @@ namespace dmGameSystem
                 value = &value[value_index];
             }
 
-            out_desc.m_ArraySize = constant_array_size;
-            out_desc.m_IsArray   = constant_array_size > 1;
+            out_desc.m_ValueType   = (constant_array_size > 1) ? dmGameObject::PROP_VALUE_ARRAY : 0;
```
Comment (JCash) [inline engine/gamesys/src/gamesys/gamesys_private.cpp:86]: We only have "variable"/"array" support in shaders.
So, shouldn't this always be PROP_VALUE_ARRAY ?

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h (hunk 18 above)
Comment (JCash) [reply engine/gameobject/src/dmsdk/gameobject/gameobject.h:261]: We generally use 0 for the default (most common) option. Which is "arrays" in our case (i.e. variables)

**Review by JCash — CHANGES_REQUESTED** [review]
Some minor changes needed

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h (hunk 18 above)
Comment (AGulev) [reply engine/gameobject/src/dmsdk/gameobject/gameobject.h:261]: We have error mesage that uses this value:

```
                if (index_requested && (property_desc.m_ValueType != dmGameObject::PROP_VALUE_ARRAY))
                {
                    return luaL_error(L, "Options table contains index, but property '%s' is not an array.", dmHashReverseSafe64(property_id));
                }
```

File: engine/gameobject/src/gameobject/gameobject_script.cpp [hunk 28]
```diff
@@ -623,18 +635,45 @@ namespace dmGameObject
         {
         case dmGameObject::PROPERTY_RESULT_OK:
             {
-                if (index_requested && !property_desc.m_IsArray)
```
Comment (AGulev) [inline engine/gameobject/src/gameobject/gameobject_script.cpp:626]: this one

File: engine/gameobject/src/dmsdk/gameobject/gameobject.h (hunk 18 above)
Comment (JCash) [reply engine/gameobject/src/dmsdk/gameobject/gameobject.h:261]: Yes, and you need to refactor such code to fit our new code?

Comment (JCash) [reply engine/gameobject/src/dmsdk/gameobject/gameobject.h:261]: E:g. what are the possible values of "property_desc.m_ValueType" ?

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 29]
```diff
@@ -3020,9 +3020,7 @@ namespace dmGameObject
         if (instance == 0)
             return PROPERTY_RESULT_INVALID_INSTANCE;
 
-        // Default array size to 0, it should not be used as an array by default
-        out_value.m_ArraySize = 0;
-        out_value.m_IsArray = 0;
+        out_value.m_ValueType = 0;
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject.cpp:3023]: Btw, since we now have an enum, we should use that value instead:

```
out_value.m_ValueType = PROPERTY_TYPE_ARRAY;
```

File: engine/gameobject/src/gameobject/gameobject_script.cpp (hunk 28 above)
Comment (JCash) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:626]: I don't see a reason that this new if-statement would fail, just because ARRAY=0 and HASHTABLE=1 ?

Comment (AGulev) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:626]: In café when it is not an array and not a hash table but user tries to specify index

Comment (AGulev) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:626]: If I keep one byte I need to add array length in this structure again, and use it instead

Comment (JCash) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:626]: >In case when it is not an array and not a hash table but user tries to specify index

But as we discussed previously, it can only be an array _or_ a hashtable.
See the discussion about the property type

Comment (JCash) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:626]:
```
    enum PropertyValueType
    {
        PROP_VALUE_HASHTABLE,
        PROP_VALUE_ARRAY    
    };
 ```

Comment (JCash) [reply engine/gameobject/src/gameobject/gameobject_script.cpp:626]: In our discussion we said that a "variable" is an array of size 1.

**Review by JCash — APPROVED** [review]
